### ISSUE TITLE
should never request a previously received block

### ISSUE THREAD

**Issue by tromp** [mimblewimble/grin#715]
In my http://tromp.github.io/grin.log, a block with id 8485fbb4 is requested in the following steps:

Feb 17 18:11:49.720 INFO Received block headers [8485fbb4, ... ]  from 51.15.77.170:13414
Feb 17 18:11:50.311 DEBG block_sync: 49734/49946 requesting blocks [8485fbb4,  ... ] from 10 peers
Feb 17 18:11:50.311 DEBG Requesting block 8485fbb4 from peer 176.9.84.75:13414.
Feb 17 18:11:51.066 DEBG handle_payload: Block 8485fbb4
Feb 17 18:11:51.066 DEBG Received block 8485fbb4 at 49735 from network, going to process.
Feb 17 18:11:51.066 DEBG pipe: process_block 8485fbb4 at 49735 with 0 inputs and 1 outputs.
Feb 17 18:11:51.069 DEBG Starting new sumtree extension.
Feb 17 18:11:51.173 DEBG chain_store: get_output_pos: Commitment(09f6d74c9da46c52afdaba5eb98cb9811d3f658c4b2f127cc7d37034bb9c88d5d7), Err(NotFoundErr)
Feb 17 18:11:51.174 DEBG sumtree: get_output_pos (idx): Commitment(09f6d74c9da46c52afdaba5eb98cb9811d3f658c4b2f127cc7d37034bb9c88d5d7), Err(NotFoundErr)
Feb 17 18:11:51.176 DEBG chain_store: get_kernel_pos: Commitment(0937d7ed18f97d3b6ce8544ad5aa18faf50e2acc388d1f3ac389aea0ff5f3cda24), Err(NotFoundErr)
Feb 17 18:11:51.176 DEBG sumtree: get_kernel_pos (idx): Commitment(0937d7ed18f97d3b6ce8544ad5aa18faf50e2acc388d1f3ac389aea0ff5f3cda24), Err(NotFoundErr)
Feb 17 18:11:51.179 DEBG -- outputs --
Feb 17 18:11:51.179 DEBG -- range proofs --
Feb 17 18:11:51.179 DEBG -- kernels --
Feb 17 18:11:51.179 DEBG validate_block: utxo roots - 9f4d2327, 9f4d2327
Feb 17 18:11:51.179 DEBG validate_block: rproof roots - c4790da8, c4790da8
Feb 17 18:11:51.179 DEBG validate_block: kernel roots - e70ec175, 791783ab
Feb 17 18:11:51.180 DEBG Error returned, discarding sumtree extension.
Feb 17 18:11:51.180 INFO Rejected block 8485fbb4 at 49735: InvalidRoot
Feb 17 18:11:51.180 DEBG Block 8485fbb4 refused by chain: InvalidRoot
Feb 17 18:11:51.180 DEBG block_received: 8485fbb4 is a bad block, resetting head
Feb 17 18:11:51.180 DEBG save_peer: V4(176.9.84.75:13414) marked Banned
Feb 17 18:11:51.180 DEBG Banning peer 176.9.84.75:13414

Ok, so there is a problem with this block. But ANOTHER problem is that it keeps on requesting this block!

Feb 17 18:12:03.195 INFO Received block headers [8485fbb4, ...]  from 204.44.91.190:13414
Feb 17 18:12:03.195 DEBG pipe: sync_block_header 8485fbb4 at 49735
Feb 17 18:12:03.198 DEBG pipe: update_header_head: 357099382, 357097839
Feb 17 18:12:03.198 DEBG header head 357099382 @ 49735 [8485fbb4]
Feb 17 18:12:03.199 DEBG sync head 357099382 @ 49735 [8485fbb4]
Feb 17 18:12:10.615 DEBG sync: locator: [27f77b10, a776fbd5, 8388729c, 707763ca, 7dd44d88, 14d37e25, dfe271aa, 14390235, 8485fbb4, d20d973a, d2cde708, fa6fcb5c, 5e3bb75e, 88b8698b, 03c1a7f1, e17ea64f]
Feb 17 18:12:10.615 DEBG sync: request_headers: asking 51.15.77.170:13414 for headers, [27f77b10, a776fbd5, 8388729c, 707763ca, 7dd44d88, 14d37e25, dfe271aa, 14390235, 8485fbb4, d20d973a, d2cde708, fa6fcb5c, 5e3bb75e, 88b8698b, 03c1a7f1, e17ea64f]
Feb 17 18:12:10.700 DEBG block_sync: 49734/50245 requesting blocks [8485fbb4, ... ] from 9 peers
Feb 17 18:12:10.700 DEBG Requesting block 8485fbb4 from peer 165.227.63.166:13414.

In fact, I had

$ grep -c "Requesting block 8485fbb4" grin.log 
44

when I decided to kill the server

I don't think receiving a block a 2nd, 3rd, ... and 44th time can possibly result in anything other than the same InvalidRoot it concluded the first time, so it's just wasting bandwidth of peers.

**Comment by ignopeverell** [comment]
Right, ideally sync would mark that entire fork as invalid so it doesn't try to follow it again with identical consequences. For now, we just ban the peer that gives us the bad block and at some point we just won't be connected with the "bad block peers".

Practically, this is the consequence of another block building or validation bug, and so I'm not sure the above logic would be very useful on a sane network (like us in a year, say).

**Comment by heunglee** [comment]
The feature will promote a healthy eco system of grin. Based on my observation and analysis of syncing problems, InvalidRoot error can be caused by local sumtree/utxo corruption from unexpected power outage, specially for individual's nodes of p2p. If corruption of file db causes InvalidRoot, then the wasteful repeats of requesting start off. It would be nicer for the admin to be notified of problem. For miners, since they might have 99.9% uptime. So for them, not necessary, but individuals running node without protection from power outage might complain like @tromp.

**Comment by mfronczyk** [comment]
Could we protect the local db from corruption on unexpected power outages? Or maybe create an automatic recovery process?

**Comment by ignopeverell** [comment]
I think we need some sort of metadata file with the storage that stores total structure sizes after all writes have succeeded. Otherwise a failure in between writes is always going to be a problem.
